Changed script to work with Elastic 6.0.0.

Zakkery opened the pull request:
Hi,

This should be working on Elastic 6.0.0. They really simplified design of scoring scripts so it is now only one file, pretty much. 
Inline scripts are now depreciated so using 'source' field instead. 
I had to also alter test files since Elastic 6.0.0 uses Netty4Plugin now and some settings are now disabled. 
I tested it on real data and it seems to work, but let me know if you have any concern.

Best,
Zakkery

Commit:
Changed script to work with Elastic 6.0.0. Inline scripts are now depreciated so using 'source' field instead

Commit:
Changed script to work with Elastic 6.0.0. Inline scripts are now depreciated so using 'source' field instead

Commit:
Changed script to work with Elastic 6.0.0. Inline scripts are now depreciated so using 'source' field instead

Zakkery committed:
renamed engine appropriately

Zakkery committed:
renamed engine appropriately

Zakkery committed:
renamed engine appropriately

lior-k commented:
Hi Zakkery.
Thanks for the PR.
I'll add my comments in-line
But 1 in general - several files instead of 1 helps me keep the clutter of Plugin boiler plate away & focus only on the 2 important methods: run & compile.
I would really appreciate it If you can move the clutter to separate files & keep indentation to a minimum

lior-k commented on src/main/java/com/liorkn/elasticsearch/plugin/VectorScoringPlugin.java:
```diff
@@ -13,23 +13,167 @@
+                                            docVectorNorm += docVector[i]*docVector[i];
+                                        }
+                                        // dot product
+                                        score += docVector[i] * searchVector.get(i).doubleValue();
```
Since this line is done <num of vectors in corpus>*<vector dimension> times, it needs to be as fast as possible.
For performance reasons searchVector should be a primitive type

lior-k commented on src/main/java/com/liorkn/elasticsearch/plugin/VectorScoringPlugin.java:
```diff
@@ -13,23 +13,167 @@
+            }
+            
+        	// we use the script "source" as the script identifier
+            if ("binary_vector_score".equals(scriptSource)) {
```
to diminish indentation please turn this if to a fast reject:
if (!"binary_vector_score".equals(scriptSource) {
  throw new IllegalArgumentException("Unknown script name " + scriptSource);
}
// no need for an else section

lior-k replied:
and please put "binary_vector_score" in a final static class member. It will minimize the amount of string allocations

ran22 committed:
optimize plugin and use float32 (#19)

Note that all vectors should be float instead of doubles
This introduces ~ 50% performance boost. so it's worth it

Commit:
updated jackson version

Commit:
added a cosine score test

Zakkery committed:
Broke it back into multiple files for simplicity, added support for encodedVector field, added runAsLong(), changed searchVector to be a primitive type, made SCRIPT_SOURCE private static class member

Zakkery committed:
Broke it back into multiple files for simplicity, added support for encodedVector field, added runAsLong(), changed searchVector to be a primitive type, made SCRIPT_SOURCE private static class member

Zakkery committed:
Broke it back into multiple files for simplicity, added support for encodedVector field, added runAsLong(), changed searchVector to be a primitive type, made SCRIPT_SOURCE private static class member

Zakkery committed:
Removed unused import

Zakkery committed:
Removed unused import

Zakkery committed:
Removed unused import

lior-k committed:
Update README.md

Fixed an issue with the go example code

Zakkery commented:
I broke it back into multiple files, like you suggested and fixed code according to your comments.

Zakkery committed:
Removed unused method

Zakkery committed:
Removed unused method

Zakkery committed:
Removed unused method

Commit:
changed a double usage in a test to float. as part of the move to floats (for performance reasons), this was missed

lior-k commented:
Hi Zakkery. note that master was changed to use float vectors instead of doubles.
please review your changes against master.
thanks

Zakkery committed:
Fixed vector to be float instead of double

Zakkery committed:
Fixed vector to be float instead of double

Zakkery committed:
Fixed vector to be float instead of double

Zakkery committed:
Fixed vector to be float instead of double

Zakkery committed:
Merge branch 'es-6.0' of https://github.com/Zakkery/fast-elasticsearch-vector-scoring into es-6.0

Zakkery commented:
> Hi Zakkery. note that master was changed to use float vectors instead of doubles.
> please review your changes against master.
> thanks

I just did a rebase, I guess I forgot to do it, now should be all floats and merged properly.

lior-k commented on src/main/java/com/liorkn/elasticsearch/script/VectorScoreScript.java:
```diff
@@ -1,114 +1,90 @@
+            
+            final int len = input.readVInt();
+            // in case vector is of different size
+            if (len != inputVector.length * Double.BYTES) {
```
this will always be true since we moved to floats.
I would like the scoring to fail if we have a problem with the vectors instead of hiding it by returning 0.
It helped us find bugs in our vector creation code. So please remove this check & the try..catch on the entire function

Zakkery replied:
I cannot remove try/catch - binaryEmbeddingReader.binaryValue() can throw an IOException:
    public abstract BytesRef binaryValue() throws IOException

lior-k commented on src/test/java/com/liorkn/elasticsearch/PluginTest.java:
```diff
@@ -71,8 +71,8 @@ public void test() throws Exception {
         final Map<String, String> params = new HashMap<>();
         params.put("refresh", "true");
         final ObjectMapper mapper = new ObjectMapper();
```
please update this file with master - it's missing some assertions at the end

Zakkery replied:
It should be the same now

Zakkery committed:
Merge branch 'es-6.0' of https://github.com/Zakkery/fast-elasticsearch-vector-scoring into es-6.0

Zakkery committed:
Fixed testing and checking of sizes
